### PR TITLE
Make CacheRecord immutable, insist on no null AccountRecords

### PR DESCRIPTION
**Pull request by AdamBJohnsonx**

### CODE REVIEW

**Comment by iambmelt** [comment]
@AdamBJohnsonx Can you link the issue driving these changes?

**Comment by AdamBJohnsonx** [comment]
> @AdamBJohnsonx Can you link the issue driving these changes?

https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/1159

**Review by iambmelt — APPROVED** [review]
I don't see any obvious issues with these changes, though I am curious how the original error in https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/1159 occurred

**Comment by AdamBJohnsonx** [comment]
> I don't see any obvious issues with these changes, though I am curious how the original error in [AzureAD/microsoft-authentication-library-for-android#1159](https://github.com/AzureAD/microsoft-authentication-library-for-android/issues/1159) occurred

I am, too.  But finding it on construction is easier than finding it on reference.

File: common/src/main/java/com/microsoft/identity/common/internal/cache/CacheRecord.java [hunk 1]
```diff
@@ -22,59 +22,55 @@
...
-    public void setAccount(final AccountRecord account) {
-        mAccount = account;
+    private final AccountRecord mAccount;
+    private final AccessTokenRecord mAccessToken;
+    private final RefreshTokenRecord mRefreshToken;
+    private final IdTokenRecord mIdToken;
+    private final IdTokenRecord mV1IdToken;
+
+    public static class CacheRecordBuilder {
+        public CacheRecordBuilder mAccount(final @NonNull AccountRecord mAccount) {
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/CacheRecord.java:47]:
```suggestion
        public CacheRecordBuilder mAccount(final @NonNull AccountRecord account) {
```

File: common/src/main/java/com/microsoft/identity/common/internal/cache/CacheRecord.java [hunk 2]
```diff
@@ -22,59 +22,55 @@
...
-        mAccount = account;
+    private final AccountRecord mAccount;
+    private final AccessTokenRecord mAccessToken;
+    private final RefreshTokenRecord mRefreshToken;
+    private final IdTokenRecord mIdToken;
+    private final IdTokenRecord mV1IdToken;
+
+    public static class CacheRecordBuilder {
+        public CacheRecordBuilder mAccount(final @NonNull AccountRecord account) {
+            if (mAccount == null) {
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/CacheRecord.java:48]:
```suggestion
            if (account == null) {
```

File: common/src/main/java/com/microsoft/identity/common/internal/cache/CacheRecord.java [hunk 3]
```diff
@@ -22,59 +22,55 @@
...
+                throw new NullPointerException("The account record for a CacheRecord may not be null");
+            }
+            this.mAccount = mAccount;
```
Comment (AdamBJohnsonx) [inline common/src/main/java/com/microsoft/identity/common/internal/cache/CacheRecord.java:51]:
```suggestion
            this.mAccount = account;
```

**Pull request merged**